test/deployframework/local.go

```diff
@@ -0,0 +1,206 @@
+package deployframework
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	ansibleRunnerPath        = "/tmp/ansible-operator/runner"
+	meteringOperatorName     = "metering-ansible-operator"
+	meteringOperatorLogName  = "metering-operator.log"
+	reportingOperatorLogName = "reporting-operator.log"
+	destKubeConfigPath       = "/kubeconfig"
+
+	enableDebug        = true
+	disableOCPFeatures = false
+)
+
+// LocalCtx holds all the necessary information to run e2e tests locally
+type LocalCtx struct {
+	Namespace                string
+	BasePath                 string
+	KubeConfigPath           string
+	MeteringOperatorImage    string
+	ReportingAPIURL          string
+	MeteringOperatorPID      int
+	ReportingOperatorPID     int
+	ReportingOperatorTLSPath string
+	Logger                   logrus.FieldLogger
+}
+
+// RunMeteringOperatorLocal is a method that runs the metering-operator locally
+func (lc *LocalCtx) RunMeteringOperatorLocal() error {
+	dockerCmd := exec.Command(
+		"docker", "run",
+		"--rm",
+		"-u", "0:0",
+		"--name", meteringOperatorContainerName,
+		"-v", lc.KubeConfigPath+":/kubeconfig",
+		"-v", ansibleRunnerPath,
+		"-e", "KUBECONFIG="+destKubeConfigPath,
+		"-e", "OPERATOR_NAME="+meteringOperatorName,
+		"-e", "POD_NAME="+meteringOperatorName,
+		"-e", "WATCH_NAMESPACE="+lc.Namespace,
+		"-e", "ENABLE_DEBUG="+strconv.FormatBool(enableDebug),
+		"-e", "DISABLE_OCP_FEATURES="+strconv.FormatBool(disableOCPFeatures),
+		lc.MeteringOperatorImage,
+	)
+
+	lc.Logger.WithFields(logrus.Fields{
+		"command": dockerCmd.Args,
+	}).Infof("Running the metering-operator docker container")
+
+	logFile, err := ioutil.TempFile(lc.BasePath, meteringOperatorLogName)
+	if err != nil {
+		return fmt.Errorf("failed to create the metering-operator container output log file: %v", err)
+	}
+
+	lc.Logger.Infof("Storing the metering-operator container logs to the '%s' path", logFile.Name())
+
+	dockerCmd.Stdout = logFile
+	dockerCmd.Stderr = logFile
+
+	err = dockerCmd.Start()
+	if err != nil {
+		return fmt.Errorf("failed to run the metering-operator container: %v", err)
+	}
+
+	lc.MeteringOperatorPID = dockerCmd.Process.Pid
```

Comment:
This is fine I think, but I wonder if we should be storing the `dockerCmd` too, so we can call `cmd.Wait()` during cleanup.

Reply:
Actually, we should definitely be storing `dockerCmd` and not `MeteringOperatorPID`. Two reasons:
- in cleanupPID you use `FindProcess` to get an `*os.Process` to kill, but `dockerCmd.Process` is already an `*os.Process` so you're unnecessarily looking up the process when we already have it.
- We want to use `cmd.Wait()`

Reply:
I was hesitant to use `cmd.Wait()` due to the reconciliation aspect, but never got around to testing the behavior of that. I think what you're proposing makes more sense for both the metering and reporting operator commands used, e.g.

```go
type LocalCtx struct {
    ...
    MeteringOperatorProcess *os.Process
    ReportingOperatorProcess *os.Process
    ...
}
```

Reply:
Hmm I don't think we want to use `cmd.Wait()` as cleanup just hangs while waiting for the reporting-operator to finish execution. I guess we just need to be more mindful in cleaning up the resources used by reporting-operator-local ourselves.

Reply:
Well I guess we could just kill the process then release the resources, e.g.
```go
err := cmd.Process.Kill()
if err != nil { ... }

err = cmd.Process.Release()
if err != nil { ... }
```

Reply:
Yeah I agree - I read up on how release is handled, and I don't think we would want to use that, especially if we can just send an interrupt to the script, wait for it to complete (which releases the all resources). This means we don't need commands to remove the docker container, or command PIDS anymore.